Try/google maps api key

jeffersonrabb opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

API for setting/updating/deleting a persisted Google Maps API key which is used by the Map block. This PR is for testing only.

~~JN: https://jurassic.ninja/create/?gutenberg&gutenpack&shortlived&jetpack-beta&calypsobranch=atavist/map-block-api-key-integration&branch=try/google-maps-api-key~~

## Testing Instructions

_IMPORTANT NOTE: These testing notes apply to the editor only. If you create a map block following these instructions and then publish or preview, nothing will show up. This is because of an unrelated change needed in Jetpack, which will be included in a new PR after this one is merged._

1) Spin up a JN instance with this link: https://jurassic.ninja/create/?gutenberg&gutenpack&shortlived&jetpack-beta&calypsobranch=master&branch=try/google-maps-api-key
2) Click `Set up Jetpack`
3) Click `Start with free`
4) Click `Activate recommeded features`
5) Navigate to `Settings->Jetpack Constants`
6) Check the box by `JETPACK_BETA_BLOCKS` and `Save Changes`
7) Create new Post
8) Click to add a block, search for `map`, add
9) Paste in a bogus Mapbox key (e.g. `notarealmapboxkey`). You should get an `Invalid API key` error message. This tests the validation code in the PUT method.
10) Paste in a real Mapbox key. You can use this key, which I'll deactivate as soon as the branch is merged: `pk.eyJ1IjoicmFiYmVyc29uIiwiYSI6ImNqb2htaWh1dTAxNnYza250dGc5N2FiamUifQ.hEAdji_oZcG0NI9VBswspw` - this should be accepted and the Map should load. This demonstrates the PUT method.
11) Click on the block, make sure Block Settings sidebar is visible, and open the `Mapbox Access Token` Panel.
12) Click `Remove Key`. This should return the block to initial state. This is the DELETE method in action.
13) Paste the real key back in.
14) Open `Mapbox Access Token` panel again, alter the token and click `Update Key`. This should show the `Invalid API Key` error and return the block to its initial state.

jeffersonrabb committed:
REST API to get and set Google Maps API Key.

jeffersonrabb committed:
REST API to get and set Google Maps API Key.

jeffersonrabb committed:
apiFetch as a dependency for view script.

jeffersonrabb committed:
Replacing get_options and update_options with Jetpack_Options::get_options / Jetpack_Options::update_options. White-listing 'googlemaps_api_key' option. Removing service check, because white-listing makes this unnecessary.

jeffersonrabb committed:
Replacing get_options and update_options with Jetpack_Options::get_options / Jetpack_Options::update_options. White-listing 'googlemaps_api_key' option. Removing service check, because white-listing makes this unnecessary.

jeffersonrabb committed:
Google Maps API key validation. Delete API. PHPcs formatting.

jeffersonrabb committed:
Google Maps API key validation. Delete API. PHPcs formatting.

enejb commented on _inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -2956,6 +2985,111 @@ private static function core_get_plugins() {
+		$address = '1+broadway+new+york+ny+usa';
+		$path    = 'https://maps.googleapis.com/maps/api/place/findplacefromtext/json?inputtype=textquery&input=' . $address;
+		$path    = add_query_arg( 'key', $key, $path );
+		$json    = json_decode( wp_remote_retrieve_body( wp_remote_get( esc_url( $path, null, null ) ) ), true );
```
`wp_remote_get( esc_url( $path, null, null ) )`  can sometimes return a wp_error object. 
So it is a good practice to check for it  before passing that the result to `wp_remote_retrieve_body( )`

```
$result = wp_remote_get( esc_url( $path, null, null ) );
if ( is_wp_error( $result ) ) {
   return array( 'status' => false, 'error_message' => $result->error_message() ) //
}
$json = json_decode( wp_remote_retrieve_body( $result ), true );
```

mdawaffe replied:
This should probably be `esc_url_raw()`. `esc_url()` is essentially: `HTML_encode( esc_url_raw() )`.

mdawaffe replied:
We should also be using `wp_safe_remote_get()`.

jeffersonrabb replied:
`wp_remote_get()` error catch, `esc_url_raw()` and `wp_safe_remote_get()` all resolved in https://github.com/Automattic/jetpack/pull/10393/commits/8274b5e211418496252ea6a58c6882d06c9c5ba9

enejb commented on _inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -2956,6 +2985,111 @@ private static function core_get_plugins() {
+	 *     @type string $slug Plugin slug with the syntax 'plugin-directory/plugin-main-file.php'.
+	 * }
+	 */
+	public static function get_api_key( $request ) {
```
This is a bit more nit picky. I think it would be cleared if we renamed `get_api_key`  to `get_service_api_key` and the same for update and delete methods.

jeffersonrabb replied:
Changed this for all related functions (`get_service_api_key` | `update_service_api_key` | `delete_service_api_key` | `validate_service_api_key` | `validate_service_api_key_googlemaps`) in https://github.com/Automattic/jetpack/pull/10393/commits/8274b5e211418496252ea6a58c6882d06c9c5ba9

enejb commented:
I have tested this with the above link and it works really well! Nicely done!

mdawaffe commented on _inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -2956,6 +2985,111 @@ private static function core_get_plugins() {
+	 */
+	public static function get_api_key( $request ) {
+		$service = $request['service'];
+		$option  = self::key_for_api_service( $service );
```
I know `Jetpack_Options` restricts what option names we can pass, but it still feels dangerous not to restrict the `service` values here (and in the other endpoints).

For example, if some future code added a `secret_api_key` option, we could get/delete/update it via the API: `jetpack/v4/api-key/secret`.

jeffersonrabb replied:
`esc_url_raw()` change made in https://github.com/Automattic/jetpack/pull/10393/commits/8274b5e211418496252ea6a58c6882d06c9c5ba9

jeffersonrabb replied:
> it still feels dangerous not to restrict the service values here (and in the other endpoints).

Very good point. I added a check in https://github.com/Automattic/jetpack/pull/10393/commits/18cfd08d0ce4c3e4030303dec539e5f5db3fdeae

mdawaffe commented on _inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -415,6 +415,35 @@ public static function register_endpoints() {
+		// Get and set API keys.
+		register_rest_route(
+			'jetpack/v4',
+			'/api-key/(?P<service>[a-z\-_]+)',
```
I suggest `/api-keys/` (plural). Even if there is no list endpoint, we're still talking about a collection (of API keys), so it should be plural (like the posts, comments, etc. endpoints).

jeffersonrabb replied:
Combined this with a comment from @enejb and opted for `/service-api-keys/` to be fully clear about the purpose of the endpoint.

mdawaffe commented on _inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -415,6 +415,35 @@ public static function register_endpoints() {
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => __CLASS__ . '::get_api_key',
+					'permission_callback' => __CLASS__ . '::view_admin_page_permission_check',
```
Echoing @enejb's [comment](https://github.com/Automattic/jetpack/commit/929cbe3ee6720fddf0704818a877779cd2de0784#r31027363), I think we need to restrict this more.

`::view_admin_page_permission_check()` just wraps a check for the `jetpack_admin_page` capability, which is just [`read`](https://github.com/Automattic/jetpack/blob/a5cb74e97db5a1df5134827eb5d28b3fb7051c87/class.jetpack.php#L917-L924).

jeffersonrabb replied:
@mdawaffe, @enejb what would your suggestions for `permission_callback` be? For the GET endpoint we need it to be entirely unrestricted, since this is accessed from the frontend by non-editors. For POST and DELETE, I'm less sure what the appropriate restriction would be?

jeffersonrabb replied:
> What's the frontend?

Sorry, I meant "any old visitors to the website". Is there a more accepted way to describe?

`:: update_settings_permission_check()` sounds good. My only question is whether there's a category of user who can see & edit the block, but cannot set or change the API key? This would be a bad situation, since the block is a brick without the key. So, if an editor adds the block for the first time but is blocked from setting the Google API key, they are stuck.

jeffersonrabb replied:
I'm hoping to avoid getting into the business of assessing the current user's access privileges unless absolutely necessary—I haven't explored it yet but it may be a little hairy to implement in these early wild west days of Gutenberg, plus it adds a new layer of testing requirements. Scenarios 1, 2, and 4 all require this level of sophistication. And I'm also hoping to avoid per-block keys (scenario 3), because that will get mega-annoying for the user if they use maps with any regularity. 

One major advantage of a site-wide API key is that it can be globally swapped if compromised, if a new entity takes over the billing, etc. Imagine a nasty little situation where a site has hundreds of maps and has to edit every one of them if for some reason their key needs to be updated. 

My $0.02: What's the argument against simply letting Authors add/update the key? It seems to me that on balance, the risk of malicious intent there is more palatable than complicating the UI for setting up the key. Also note that this key is currently used for nothing but the Map block, so if it gets inadvertently messed up, the risk isn't all that dire. If you feel differently, though, I'm happy to dig in! 

Also, another note: my eventual hope is that in future iterations of the API, Automattic provides a centralized key which can be used only in the editor and preview. This will allow any JP customer to use the block immediately—they will only have to add a key when they are ready to begin publishing. When this level of sophistication becomes reality might be a good time to tighten up the permissions, since we will no longer be as concerned about the risks of a bad first experience.

mdawaffe replied:
As it stands, anyone with an account (even people with the "Subscriber" role) can add/edit/delete the key. Imagine a large site that's set up to encourage frequent commenters to sign up for the subscriber role (see wp-admin -> Settings -> General -> "Anyone can register").

Even for Authors (i.e., users with the "Author" role), this still seems odd; in general, Authors in WordPress are allowed to do things to their posts, but are not allowed to influence others' posts or site-wide settings. Suppose an Author, before leaving an organization's site, updates the key to one they control. Now an ex-employee (or whatever) has access to stats of the site and can even break the site by invalidating the key.

Yes, this sort of thing would still be a problem with an Admin or Editor leaving, but organizations know to audit things when users of that role leave. Since WordPress is set up to limit the access Authors have, though, most organizations wouldn't think much about Authors leaving.

This is all related to the block visibility discussion at p1HpG7-5JG-p2.

Perhaps a compromise?:
* Subscribers and Contributors can not write to this API at all
* Authors and Editors can add a key but not edit or delete one
* Admins can add, edit, or delete.

(Or perhaps group Editors and Admins together?)

Also, I believe it's the case that the vast majority of sites only have one user account. I don't know if the vast majority of "interesting" (higher traffic, paid customers, publishing volume, maps block users, whatever metric you pick) sites have only one user account, though.

jeffersonrabb replied:
Thinking about this a little more I don't think there's a meaningful difference between CREATE and UPDATE here since the API keys are stored as site options with only one value per site. There isn't really a difference between unset and blank. So, instead of separating the two actions I adopted your stricter suggestion for UPDATE/DELETE (editors/admins only), and removed all restrictions on GET. I'll build some checking/messaging into the block so that < Editors understand why they are unable to set the key. Changes are in https://github.com/Automattic/jetpack/pull/10393/commits/516c7878dcb8dbbdfe0972a1f5c735abdff8331b.

jeffersonrabb committed:
Change to the endpoint (now service-api-keys). Renamed associated functions to include "service_", for clarity. Some security and error-catching improvements to Google Maps Key validation function.

jeffersonrabb committed:
Change to the endpoint (now service-api-keys). Renamed associated functions to include "service_", for clarity. Some security and error-catching improvements to Google Maps Key validation function.

jeffersonrabb committed:
Service validation, for additional security.

jeffersonrabb committed:
Service validation, for additional security.

jeffersonrabb committed:
Standardization of the responses from /service-api-keys/ endpoints.

jeffersonrabb committed:
Standardization of the responses from /service-api-keys/ endpoints.

jeffersonrabb committed:
Merge branch 'master' into try/google-maps-api-key

jeffersonrabb committed:
Support for Mapbox keys.

jeffersonrabb committed:
Support for Mapbox keys.

jeffersonrabb committed:
Comment fix.

jeffersonrabb committed:
Comment fix.

jeffersonrabb committed:
Mapbox API key validation.

jeffersonrabb committed:
Mapbox API key validation.

mdawaffe commented on _inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -415,6 +415,35 @@ public static function register_endpoints() {
+		// Get and set API keys.
+		register_rest_route(
+			'jetpack/v4',
+			'/service-api-keys/(?P<service>[a-z\-_]+)',
```
When I need to have a hyphen in a regular expression, I put it at the end:
`(?P<service>[a-z_-]+)`

Otherwise, I have to think about whether it's a range operator (from slash to underscore) or an escaped hyphen.

The above isn't advice :) Just a bikeshed about what makes sense to me.

jeffersonrabb committed:
Stricter permissions for service api keys UPDATE/DELETE. User must be editor or administrator. No check at all for GET.

jeffersonrabb committed:
Stricter permissions for service api keys UPDATE/DELETE. User must be editor or administrator. No check at all for GET.

jeffersonrabb committed:
Removing support for Google Maps key, as Map block has been rewritten using Mapbox.

jeffersonrabb committed:
Removing support for Google Maps key, as Map block has been rewritten using Mapbox.

jeffersonrabb commented on _inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -2956,6 +2997,182 @@ private static function core_get_plugins() {
+		Jetpack_Options::delete_option( $option );
+		$message = esc_html__( 'API key deleted successfully.', 'jetpack' );
+		return array(
+			'code'            => 'success',
```
The status is unused, but in the Map block I am displaying the message to the user when errors occur.

jeffersonrabb commented:
> We intend to launch this block on WordPress.com as well as in Jetpack, true?

Unclear. It has been discussed but nothing definitive. If I were planning for that outcome, are you saying I should change the namespace to `wpcom/v2` in the Jetpack codebase? Or is it more complicated than that?

jeffersonrabb committed:
Merge branch 'master' into try/google-maps-api-key

jeffersonrabb committed:
Merge branch 'master' into try/google-maps-api-key

jeffersonrabb committed:
Merge branch 'master' into try/google-maps-api-key

jeffersonrabb committed:
Comment explaining the omission of permission callback on the GET method.

jeffersonrabb committed:
Comment explaining the omission of permission callback on the GET method.

jeffersonrabb committed:
Merge branch 'master' into try/google-maps-api-key

simison committed:
Revert "Simple payments widget: remove Indian Rupees (#10581)" (#10610)

This reverts commit 99f9b5adb17ba47db8653e64be4c563e1d342cea.

jeherve commented:
Any chance you could add some testing steps to the PR, so I know what functionality  and use cases I should be testing to make sure this works?

Thank you!

jeffersonrabb commented:
> Any chance you could add some testing steps to the PR, so I know what functionality and use cases I should be testing to make sure this works?

Certainly! Just added comprehensive JN testing instructions fo the description.

georgestephanis committed:
Calypsoify: Refactor and add autoupdate functionality (#10371)

Iterations on the current pass.  Reorganize some assets, tidy up some formatting and code patterns, convert the css to scss to make for easier maintenance going forward and abstract out the color codes to allow for semantic renaming in a future pass.

Oh, and add autoupdate management functionality to the plugins page.

![pageshot of plugins test local wordpress thingy wordpress 2018-11-09-0843 24](https://user-images.githubusercontent.com/941023/48265919-d8e54280-e3fb-11e8-96db-3b789a7a502c.png)

Fixes #10474

#### Changes proposed in this Pull Request:

* Improve Calypso integration for relevant sites, and add admin ui autoupdate functionality.

#### Testing instructions:

* Install some plugins that are not in the repository (you should be able to select some from jurassic.ninja/specialops)
* Open your dev console and keep an eye out for js errors 
* Visit `/wp-admin/plugins.php?calypsoify=1`
* Test out the autoupdate plugin toggles
* Test out the bulk-select of auto-updates 
* Visit `/wp-admin/` to clear calypsoify skin.

mdawaffe committed:
REST API: Add Field Controller (#10600)

* REST API: Add Field Controller

Just as `WP_REST_Controller` is mostly just a fancy wrapper for
`register_rest_route()`, `WPCOM_REST_API_V2_Field_Controller` is a wrapper
for `register_rest_field()`. It handles a few conveniences:
* Returning the correct things (errors, default values, etc.) for
  permission check errors.
* Ensuring output is always correctly typecast.
* `?context=...` filtering for nested structures, which core does not
  handle correctly.

jeherve reviewed (dismissed):
This looks good. I only have 2 remarks, but that may more relate to the block itself:

2. I see the following error in the console:

```
mapbox-gl.js:28 Error: The user aborted a request.
    at eval (mapbox-gl.js:28)
```

<img width="692" alt="screenshot 2018-11-14 at 18 34 02" src="https://user-images.githubusercontent.com/426388/48521846-e9eae500-e83b-11e8-986f-002323efc31b.png">

2. If I have a valid key, and try to update my key to an invalid one, I would expect to receive an error that my key is invalid, but my existing key should not be deleted for me since it is still valid.

jeffersonrabb commented:
@jeherve both are good points which I'll resolve, but the resolution for both will be in `wp-calypso`. 

I believe the first error has something to do with Mapbox beginning to load the map before the library is fully ready, although I'm not positive of this. I'll run it down.

For the second (updating to an invalid API key), the Jetpack API actually works properly although the UI leads you to believe it did not. If you update the key with an invalid one, the block returns to initial state which appears as if the invalid API key was saved. However, if you refresh the page you'll find that the last valid API is still saved, and the map should load correctly. So, the fix here is in the UI, not the API, which I'll address in `wp-calypso`.

roccotripaldi committed:
Update/gutenberg block registration logic (#10625)

* update registration of markdown gutenberg block so that we can share with wpcom

* update registration of related posts gutenberg block so we can share with wpcom

jeffersonrabb committed:
Merge branch 'try/google-maps-api-key' of github.com:Automattic/jetpack into try/google-maps-api-key

jeherve reviewed (approved):
Looking good. 🚢
